### PR TITLE
fix(builtin-function): fix string conversion of days-times-durations

### LINKED ISSUE #598: Regression in string conversion of zero-length duration
**Describe the bug**
There is a regression in 1.15.3 where the string conversion of a zero-length duration leads to a non-ISO 8601 representation: `P` instead of `PT0S`.

**To Reproduce**

```
===== FEEL Engine REPL (1.15.3) ======
> Evaluate FEEL expressions using 'feel("1 + 3")'
> Evaluate FEEL unary-tests using 'unaryTests("[2..5]", 4)'
> Provide variables as Map using 'feel("x + 3", Map("x" -> 2))'
> Provide variables as JSON string using 'feel("x + 3", "{ \"x\" : 2 }")'
Welcome to the Ammonite Repl 2.5.6 (Scala 3.2.1 Java 19.0.2)
@ feel("string(duration(\"P0D\"))")
> P

===== FEEL Engine REPL (1.15.2) ======
> Evaluate FEEL expressions using 'feel("1 + 3")'
> Evaluate FEEL unary-tests using 'unaryTests("[2..5]", 4)'
> Provide variables as Map using 'feel("x + 3", Map("x" -> 2))'
> Provide variables as JSON string using 'feel("x + 3", "{ \"x\" : 2 }")'
Welcome to the Ammonite Repl 2.5.6 (Scala 3.2.1 Java 19.0.2)
@ feel("string(duration(\"P0D\"))")
> PT0S
```

**Expected behavior**
The result should be `P0D`.

_EDIT: We changed the expectation from `PT0S` to `P0D` because of a DMN TCK test case. (see https://github.com/camunda/feel-scala/issues/598#issuecomment-1527173279)_

**Environment**
* FEEL engine version: 1.15.3
* Affects: 
  * Camunda Platform 8: >8.1.7 https://github.com/camunda/zeebe/issues/11967

### LINKED ISSUE #599: Regression in string conversion of negative duration
**Describe the bug**
A negative duration is converted to a positive duration (double negation).

**To Reproduce**
```
===== FEEL Engine REPL (1.15.3) ======
> Evaluate FEEL expressions using 'feel("1 + 3")'
> Evaluate FEEL unary-tests using 'unaryTests("[2..5]", 4)'
> Provide variables as Map using 'feel("x + 3", Map("x" -> 2))'
> Provide variables as JSON string using 'feel("x + 3", "{ \"x\" : 2 }")'
Welcome to the Ammonite Repl 2.5.6 (Scala 3.2.1 Java 19.0.2)
@ feel("string(duration(\"-PT1S\"))")
> -P-1S

===== FEEL Engine REPL (1.15.2) ======
> Evaluate FEEL expressions using 'feel("1 + 3")'
> Evaluate FEEL unary-tests using 'unaryTests("[2..5]", 4)'
> Provide variables as Map using 'feel("x + 3", Map("x" -> 2))'
> Provide variables as JSON string using 'feel("x + 3", "{ \"x\" : 2 }")'
Welcome to the Ammonite Repl 2.5.6 (Scala 3.2.1 Java 19.0.2)
@ feel("string(duration(\"-PT1S\"))")
> PT-1S
```

**Expected behavior**
The result of `string(duration("-PT1S"))` should be `"-PT1S"`

**Environment**
* FEEL engine version: 1.15.3
* Affects: 
  * Camunda Platform 8: >8.1.7

### PR DESCRIPTION
**Pull request by sebi87**
* fix toString() of ValDayTimeDuration
* add test cases for negative day/time durations

closes #599 
closes #598

### CODE REVIEW

**Comment by saig0** [comment]
@sebi87 thank you for your contribution. :tada: I'll have a look at your changes soon. :eyes:

File: src/test/scala/org/camunda/feel/impl/builtin/BuiltinConversionFunctionsTest.scala [hunk 1]
```diff
@@ -211,7 +211,13 @@ class BuiltinConversionFunctionsTest
     eval(""" string(date and time("2012-12-25T11:00:00+02:00")) """) should be(
       ValString("2012-12-25T11:00:00+02:00"))
   }
+  it should "convert negative days-time-duration" in {
 
+    eval(""" string(@"-PT1S") """) should be(ValString("PT-1S"))
+    eval(""" string(@"-PT1H") """) should be(ValString("PT-1H"))
+    eval(""" string(@"-PT2M30S") """) should be(ValString("PT-2M-30S"))
+    eval(""" string(@"-P1DT2H3M4S") """) should be(ValString("P-1DT-2H-3M-4S"))
```
Comment (saig0) [inline src/test/scala/org/camunda/feel/impl/builtin/BuiltinConversionFunctionsTest.scala:219]: :x: I had a deeper look at the DMN spec. The expected format is `"-PT1S"` instead of `"PT-1S"`. 

Please see the details [here](https://github.com/camunda/feel-scala/issues/599#issuecomment-1522756437).

**Review by saig0 — CHANGES_REQUESTED** [review]
@sebi87 your changes look great. :tada: 

However, I checked the DMN spec again and the expected format is different. Please see my comment.

**Comment by sebi87** [comment]
@saig0 this should also fix #598. Isn't there the same problem with years-months-duration? "-P1Y" converts to "P-1Y" instead of "-P1Y".

File: src/test/scala/org/camunda/feel/impl/builtin/BuiltinConversionFunctionsTest.scala [hunk 2]
```diff
@@ -211,7 +211,21 @@ class BuiltinConversionFunctionsTest
...
       ValString("2012-12-25T11:00:00+02:00"))
   }
+  it should "convert zero-length days-time-duration" in {
+    eval(""" string(@"-PT0S") """) should be(ValString("PT0S"))
+    eval(""" string(@"P0D") """) should be(ValString("PT0S"))
+    eval(""" string(@"PT0H") """) should be(ValString("PT0S"))
+    eval(""" string(@"PT0H0M") """) should be(ValString("PT0S"))
+    eval(""" string(@"PT0H0M0S") """) should be(ValString("PT0S"))
+    eval(""" string(@"P0DT0H0M0S") """) should be(ValString("PT0S"))
+  }
```
Comment (saig0) [inline src/test/scala/org/camunda/feel/impl/builtin/BuiltinConversionFunctionsTest.scala:221]: :x: Actually, the expected value of a zero-length duration is `P0D`. :sweat_smile: 

I added a comment [on the issue](https://github.com/camunda/feel-scala/issues/598#issuecomment-1527173279).

File: src/main/scala/org/camunda/feel/syntaxtree/Val.scala [hunk 3]
```diff
@@ -179,21 +179,34 @@ case class ValYearMonthDuration(value: YearMonthDuration) extends Val {
...
+    signedNonZero
+      .map(a => (
+        a._1,
+        appendIfNotZero(a._2, "D"),
+        a._3,
+        appendIfNotZero(a._4, "H"),
+        appendIfNotZero(a._5, "M"),
+        appendIfNotZero(a._6, "S")
+      ).productIterator.mkString)
+      .getOrElse("PT0S")
```
Comment (saig0) [inline src/main/scala/org/camunda/feel/syntaxtree/Val.scala:209]: :x: The code seems to work. :tada: 

But it looks a bit too complicated. :dizzy_face: Please try to simplify the code. To goal is that it is easy to understand and maintain, following the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) _("Keep it simple, stupid!")_.

**Review by saig0 — CHANGES_REQUESTED** [review]
@sebi87 thank you for your fast response. :rocket: 

I have a few comments on the zero-length fix and the code style. Please have a look. 

Btw: the Clirr API check is failing at the moment. It's a false-positive. You can ignore it.

**Comment by saig0** [comment]
> Isn't there the same problem with years-months-duration? "-P1Y" converts to "P-1Y" instead of "-P1Y".

Yes, good point. :+1: We have the same issue for the years months duration.

File: src/test/scala/org/camunda/feel/impl/builtin/BuiltinConversionFunctionsTest.scala (hunk 2 above)
Comment (sebi87) [reply src/test/scala/org/camunda/feel/impl/builtin/BuiltinConversionFunctionsTest.scala:221]: 🙈 Yes, I think I can handle that. 😅

File: src/main/scala/org/camunda/feel/syntaxtree/Val.scala (hunk 3 above)
Comment (sebi87) [reply src/main/scala/org/camunda/feel/syntaxtree/Val.scala]: Yes, you're absolutely right. I refactored it and now it's more obvious I think.

**Review by saig0 — APPROVED** [review]
@sebi87 your changes look great. :tada: The PR is ready to merge. :rocket: 

Thank you for your contribution. :cake:
